src/core/logging/logger.hpp

```diff
@@ -324,14 +329,16 @@
 
 #endif
 
-#define log_and_throw(message)                                      \
-  do {                                                              \
-    auto throw_error = [&]() GL_COLD_NOINLINE_ERROR {  \
-      logstream(LOG_ERROR) << (message) << std::endl;               \
-      throw(std::string(message));                                  \
-    };                                                              \
-    throw_error();                                                  \
-  } while(0)
+#ifdef NDEBUG
+
+#define log_and_throw(message)                        \
+  do {                                                \
+    auto throw_error = [&]() GL_COLD_NOINLINE_ERROR { \
+      logstream(LOG_ERROR) << (message) << std::endl; \
+      throw(std::runtime_error(message));             \
```

Comment:
I would leave this as `std::string` for now, as per your comment re: ~20 places that would need to be updated. We could take that change as a separate PR in the future (changing it to runtime_error here, and updating all 20 places in the same commit), but for now I would just open a GitHub issue to track it (if there is not one already), and get this PR in first.

Reply:
we have a bunch of places catching by `const char*` and `const std::string&`, which is kind of confusing to me at first. That's probably also why there are lots of `catch (...)` which doesn't really help and may get things worse because we are tend to be lazy without trying to exhaust handling all possible exception types. 

20 places are not that many. I think they are side effects of this PR and we should include those in this PR so that I can run all unit tests and make sure I don't break anything.

Reply:
It would be great to do that as a separate PR, since the two work items are essentially unrelated except that they happen to touch some of the same parts of the code:

1. Get filename and line number logging on log_and_throw
2. Have log_and_throw use std::runtime_error as its exception type

These should be two separate PRs, since it's easier to review and comment on them separately, and there are various situations where this makes sense in general:

* If one PR should go in, and the other shouldn't, it's harder to separate them after the fact than to keep them separate to begin with.
* If one PR needs feedback and changes, and the other is ready to go in, they aren't both delayed.
* Reviewers who are best suited to review and discuss one of them may not be the best reviewers for the other.

Reply:
That's valuable advice. I can separate them now.